numpy/core/src/multiarray/lowlevel_strided_loops.c.src

```diff
@@ -1823,3 +1825,193 @@ mapiter_@name@(PyArrayMapIterObject *mit)
 }
 
 /**end repeat**/
+
+
+
+#define ptr_assignment1(ptr1, val1, stride1) *ptr1 = val1; ptr1 += stride1; 
+#define ptr_assignment2(ptr1, val1, stride1, ptr2, val2, stride2) ptr_assignment1(ptr1, val1, stride1) *ptr2 = val2; ptr2 += stride2; 
+
+#define loop_ptr_assignment(start, end, ptr_assignment) \
+    for (npy_intp i=start; i<end; ++i) { \
+        ptr_assignment \
+    }
+
+/**begin repeat
+ * 
+ * #dtype = npy_bool, npy_byte, npy_byte, npy_uint16, npy_int16, npy_uint32, npy_int32, npy_uint64, npy_int64, npy_float, npy_double#
+ * #name = bool, u8, i8, u16, i16, u32, i32, u64, i64, f32, f64#
+ */
+
+/**begin repeat1
+ *
+ * #layout=C,F#
+ *
+ */
+
+void nonzero_idxs_1D_@layout@_@name@(@dtype@ * data, npy_intp* idxs, npy_intp* shape, npy_intp* strides, npy_intp nonzero_count) {
+    npy_intp stride = strides[0]; 
+    npy_intp added_count = 0; 
+    npy_intp a = 0; 
+    while (added_count < nonzero_count) { 
+        *idxs = a;  
+        bool to_increment = ((bool) *data); 
+        idxs += ((int) to_increment); 
+        added_count += ((int) to_increment); 
+        data = (@dtype@ *) (((char*) data) + stride); 
+        ++a;
+    }
+}
+
+ /**end repeat1**/
+
+void nonzero_idxs_2D_C_@name@(@dtype@ * data, npy_intp* idxs, npy_intp* shape, npy_intp* strides, npy_intp nonzero_count) { 
+    npy_intp idxs_stride = 2; 
+    npy_intp size_1 = shape[1]; 
+    npy_intp stride_1 = strides[1]; 
+    npy_intp* idxs_0 = idxs;
+    npy_intp* idxs_1 = idxs + 1;    
+    npy_intp added_count = 0; 
+    npy_intp a = 0; 
+    while (added_count < nonzero_count) { 
+        npy_intp old_added_count = added_count; 
+        npy_intp b = 0;
+        while (added_count < nonzero_count && b<size_1) { 
```

Comment:
Wouldn't it be nicer to move this into the inner-most loop as a `if (added_count == nonzero_count) { return;}`?  Maybe also in the 1-D case.

Probably you can also turn the `while` loops into `for (a = 0;; a++)` which is a bit funny, but probably just as micro-optimized and easier to read?

(For the outer-most loop you can skip the `a < size_0`.

Reply:
(also here: please fix the spaces. `b < size_1`. We also normally write pointer data as `npy_intp *idx`, not `npy_intp* idx` or `npy_intp * idx`.  I don't know if thats the best style, but its consistent style :).